### ISSUE TITLE
Custom Error Types

### ISSUE THREAD

**Issue by Dylan-DPC** [sodiumoxide/sodiumoxide#221]
Currently in the crate we are doing `Err(())`. However, this isn't verbose enough for the user and he wouldn't know what error has occurred without going through the crate source. Instead we can create custom error types for different types of errors which along with a message, gives a clear meaning to the user about the error. 

A general practice is to use the [failure](https://github.com/rust-lang-nursery/failure) crate.

**Comment by nvesely** [comment]
> A general practice is to use the failure crate.

I've never used this crate, but just skimmed the README to look for no_std compatibility and saw

> failure is no_std compatible, though some aspects of it (primarily the Error type) will not be available in no_std mode.

So we'd have to do something like

```rust
    if cfg!(feature = "std") {
        return failure::Error();
    } else {
        return Err(());
    }
```

universally. Depending on what type of error-handling we want to implement, this could become more complicated.

**Comment by nvesely** [comment]
Wait that's not going to work. We'll have to write each function twice because the return types are different. Maybe this means we'll want to go a simpler route for error-handling.

**Comment by Dylan-DPC** [comment]
@nvesely ah yes forgot about the no_std part. We will keep this pending for a while till we find a good solution for this

**Comment by dnaq** [comment]
Can't we just create our own custom error type? Users of our crate can wrap it in failure if needed or do whatever they want with it.

**Comment by naftulikay** [comment]
Yes, totally agree with @dnaq

**Comment by nvesely** [comment]
Third for keeping it simple. Don't think we need to pull in additional crates for this one.

**Comment by Dylan-DPC** [comment]
I agree we may not need a crate right now for this. But failure is there for certain reasons, so in future there is a good chance we may be needing it. Can skip for now though.

**Comment by caass** [comment]
Hello! Is there any ongoing work on this issue? If not, I'd like to give it a try.

For an example use case, I'm trying to use this library with [juniper](https://github.com/graphql-rust/juniper), which will perform [automatic error conversion](https://graphql-rust.github.io/juniper/master/types/objects/error_handling.html#field-results) with anything that implements `std::fmt::Display`.

Creating a custom error type that implements `Display` would make my code a lot more ergonomic!

**Comment by mso42** [comment]
Created a PR with a sketch of a pattern for custom error types.
